File: src/functionalize.cpp
#include "functionalize.hpp"

namespace chemutils{

ORDER_DICT findHIds(std::shared_ptr<RDKit::ROMol> mol, bool unique){
    /*
    Find the hydrogens from the molecule
    If unique flag is set only symmetrically unique hydrogens will be returned
    Return a map with the format {atom order : {H index, Neighbor Atom index}}
    */
    std::vector<unsigned int> rank;
    // find symmetrically equivalent hydrogens by canonical order
    // if unique, no equivalent order will be given (unique rank for equivalent Hs)
    bool breakTies = !unique;
    RDKit::Canon::rankMolAtoms(*mol, rank, breakTies);
    ORDER_DICT orders;
    for (std::size_t j=0; j < rank.size(); j++){
        const RDKit::Atom *atom = mol->getAtomWithIdx( j );
        
        if (atom->getSymbol() == "H"){
            // Find the neighbor atom's id
            // There is only one neighbor exists for Hydrogen
            RDKit::ROMol::ADJ_ITER_PAIR id_pair = mol->getAtomNeighbors(atom);
            // Insert rank of the hydrogen and the atom id pair in map,
            // id of Hydrogen and its neighbor atom matching the sdf idx (H_id, Nbr_id).
            // Only unique rank of hydrogen is allowed as key 
            // Will produce geometrically equivalent hydrogens if unique is false 
            orders.insert({rank[j], {j, *id_pair.first}});
        }
    }
    return orders;
}

void setTetherIdx(std::shared_ptr<RDKit::ROMol> mol){
    // Set the tethering atom index on the reference lig
    // Mol props set here will be preserved when new is called for constructing 
    // the new ligands
    std::shared_ptr<RDKit::ROMol> mol2( RDKit::MolOps::removeHs( *mol ));
    // The atom index number to be tethered will be all atoms except hydrogen
    int n_atoms = mol2->getNumAtoms();
    // Construct a string of idx to be set on the mol prop
    std::string tether_idx="1";
    for (int i = 2; i <= n_atoms; i++){
        tether_idx += ", ";
        tether_idx += std::to_string(i);
    }
    // set prop for both versions of rdock
    mol->setProp("rxdock.tethered_atoms", tether_idx);
    mol->setProp("rdock.tethered_atoms", tether_idx);
}

std::shared_ptr<RDKit::ROMol> connectMols(std::shared_ptr<RDKit::ROMol> lig, std::shared_ptr<RDKit::ROMol> frag, 
    int id_ligH, int id_ligNbr, int id_fragH, int id_fragNbr)
    /*
    Connect two mols whose hydrogens are aligned into position
    Return a combined mol
    Got the idea from the python script linked below
    https://github.com/molecularsets/moses/blob/master/moses/baselines/combinatorial.py
    */
    {
        // make editable copied of the molecules
        std::shared_ptr<RDKit::RWMol> rw_lig( new RDKit::RWMol( *lig ));
        std::shared_ptr<RDKit::RWMol> rw_frag( new RDKit::RWMol( *frag ));
        // map the atom to be aligned from the two mols
        const RDKit::MatchVectType id_map = {{id_fragH, id_ligNbr}, {id_fragNbr, id_ligH}};
        // mol1->setProp( "rxdock.tethered_atoms", "cyclobutane" );
        // frag first and lig
        double _rmsd = RDKit::MolAlign::alignMol(*rw_frag, *rw_lig, -1, -1, &id_map);
        // combine the two mols
        rw_lig->insertMol( *rw_frag );
        rw_lig->addBond( id_ligNbr, lig->getNumAtoms() + id_fragNbr, RDKit::Bond::SINGLE );
        // remove the hydrogens from the ligand and the fragment
        rw_lig->removeAtom( id_fragH + lig->getNumAtoms() );
        rw_lig->removeAtom( id_ligH );
        // sanitize the combined mol 
        try{
            RDKit::MolOps::sanitizeMol(*rw_lig);
        } catch (RDKit::MolSanitizeException &e){
            std::cerr << "Fail to sanitize: " << MolToSmiles(*rw_lig) << std::endl;
            std::cerr << e.what() << std::endl;
        }
        
        return std::shared_ptr<RDKit::ROMol>( rw_lig );
}

MOL_STDSPTR_VECT functionalizeH(std::shared_ptr<RDKit::ROMol> lig,  
    std::shared_ptr<RDKit::ROMol> frag, bool unique, bool tether)
    /*
    Functionalize the ligand with a fragment on all unique positions of hydrogen atoms
    If unqiue flag is set to False, all hydrogen positions on the ligand will be used
    for forming the bonds. 
    Return a vector of ptrs of functionalized ligand
    */
    // For a single lig-frag pair generation
    {
        if (tether == true){
            setTetherIdx(lig);
        }
        ORDER_DICT orders_lig = findHIds(lig, unique);
        // only unique hydrogens position will be used on the fragament
        ORDER_DICT orders_frag = findHIds(frag, true);
        
        MOL_STDSPTR_VECT cur_combined;
        for (ORDER_DICT::iterator it=orders_lig.begin(); it!=orders_lig.end(); ++it){
            for (ORDER_DICT::iterator jt=orders_frag.begin(); jt!=orders_frag.end(); ++jt){
                cur_combined.push_back( connectMols(lig, frag, 
                it->second.first, it->second.second, 
                jt->second.first, jt->second.second) );
            }
        }
        return cur_combined;
}

COMBINED_VECT_2D functionalize_lig(std::shared_ptr<RDKit::ROMol> lig,  
    MOL_STDSPTR_VECT  frags, bool unique, bool tether)
    /*
    Enumerated functionalized molecules for a single ligand ptrs 
    and a vector of fragment ptrs
    Return a 2D vector of shared_ptrs with format 
    {lig1-frag1, lig1-frag2, ..., lig1-fragN}
    where each lig-frag pair is a vector of ptrs to functionalized ligands
    */
    {
        if (tether == true){
            setTetherIdx(lig);
        }
        // ligs H ids are determined here to avoid repition
        ORDER_DICT orders_lig = findHIds(lig, unique);
        COMBINED_VECT_2D lig_combined;
        for (std::shared_ptr<RDKit::ROMol> frag: frags){
            ORDER_DICT orders_frag = findHIds(frag, true);
            MOL_STDSPTR_VECT cur_combined;
            for (ORDER_DICT::iterator it=orders_lig.begin(); it!=orders_lig.end(); ++it){
                for (ORDER_DICT::iterator jt=orders_frag.begin(); jt!=orders_frag.end(); ++jt){
                    cur_combined.push_back( connectMols(lig, frag, 
                    it->second.first, it->second.second, 
                    jt->second.first, jt->second.second) );
                }
            }
            lig_combined.push_back(cur_combined);
        }
        
    return lig_combined;
}

COMBINED_VECT_3D combinatorialFunc(MOL_STDSPTR_VECT ligs, MOL_STDSPTR_VECT frags,
     bool unique, bool tether)
    /*
    Enumerated all the functionalized molecules 
    from a vector of ligand ptrs and a vector of fragment ptrs
    Return a vector of vectors of shared_ptrs with format 
    {{lig1-frag1, lig1-frag2, ..., lig1-fragN},
     {lig2-frag1, lig2-frag2, ..., lig2-fragN},
      ...
     {ligN-frag1, ligN-frag2, ..., ligN-fragN}}
    */ 
    {
    int is = ligs.size();
    COMBINED_VECT_3D all_combined;
    FUTURE_VECT fs(is);

    for (int i=0; i<is; i++){
        // multithreading: each thread takes on ligand and functionalize with all fragments
        fs[i] = std::async(functionalize_lig, ligs[i], frags, unique, tether);
    }

    for (int i=0; i<is; i++){
        // collect results
        // threads are async, but the results are placed acording to the ligands ids
        all_combined.push_back(fs[i].get());
    }

    return all_combined;
}

void _sdfCombinatorialWrapper(int lig_idx, std::shared_ptr<RDKit::ROMol> lig, MOL_STDSPTR_VECT frags, 
    std::string out_path, bool unique, bool tether)
    /*
    Wrapper function to for better multithreading on sdfCombinatorialFunc()
    */
    {
    if (tether == true){
        setTetherIdx(lig);
    }
    ORDER_DICT orders_lig = findHIds(lig, unique);
    int frag_idx = 0;

    for (std::shared_ptr<RDKit::ROMol> frag: frags){
        // only unique hydrogens position will be used on the fragament
        
        ORDER_DICT orders_frag = findHIds(frag, true);
        std::string fname = "combined_"+std::to_string(lig_idx)+"_"+std::to_string(frag_idx)+".sdf";
        std::string fileout = out_path+"/"+fname;
        RDKit::SDWriter writer(fileout);

        for (ORDER_DICT::iterator it=orders_lig.begin(); it!=orders_lig.end(); ++it){
            for (ORDER_DICT::iterator jt=orders_frag.begin(); jt!=orders_frag.end(); ++jt){
                writer.write( *connectMols(lig, frag, 
                    // int id_ligH, int id_ligNbr, 
                    // int id_fragH, int id_fragNbr
                    it->second.first, it->second.second, 
                    jt->second.first, jt->second.second)
                );
            }
        }
        frag_idx++;
    }
}

void sdfCombinatorialFunc(std::string lig_file, std::string frag_file, std::string out_path, 
    bool unique, bool tether)
    /*
    Enumerated all combinatorials of ligands and fragments for various H position for functionalization  
    Takes input from .sdf ligand file and .sdf fragment file 
    Files can contain multiple ligands or fragments to be connected
    Save the fused molecules in sdf files with presevation of ligands 3D coordinates
    */
    {
    if (std::filesystem::exists(out_path)){
        bool takeOwnership = true;
        bool removeHs = false;

        // read ligand mols from file
        MOL_STDSPTR_VECT ligs;
        RDKit::SDMolSupplier lig_supplier( lig_file , takeOwnership , removeHs );
        while( !lig_supplier.atEnd() ) {
            std::shared_ptr<RDKit::ROMol> lig( lig_supplier.next() );
            if( lig ) {
                ligs.push_back( lig );
            }
        }

        // read fragment mols from file
        MOL_STDSPTR_VECT frags;
        RDKit::SDMolSupplier frag_supplier( frag_file , takeOwnership , removeHs );
        while( !frag_supplier.atEnd() ) {
            std::shared_ptr<RDKit::ROMol> frag( frag_supplier.next() );
            if( frag ) {
                frags.push_back( frag );
            }
        }
        // execute with multithreaded job
        int is = ligs.size();
        FUTURE_WRAPPER_VECT fs(is);
        for (int i=0; i<is; i++){
            fs[i] = std::async(_sdfCombinatorialWrapper, i, ligs[i], frags, out_path, unique, tether);
        }
        // wait for jobs to finish
        for (int i=0; i<is; i++){
            fs[i].wait();
        }
        std::cout << ligs.size()*frags.size() << " Ligand-Fragament pairs generated." << std::endl;
    } else {
        std::cerr << "[Aborted] File path not exists!" << std::endl;
    }
}
} // namespace chemutils

File: chemutils/functionalize.py
## Frank Lab 2021
## University of Michigan

from rdkit import Chem
from rdkit.Chem.rdMolAlign import AlignMol
import copy
from itertools import product, repeat
from tqdm.contrib.concurrent import process_map
import os
from filterProp import PropFilter

def findHIds(mol, unique: bool = True):
    '''
    Find the hydrogens from the molecule
    If unique flag is set only symmetrically unique hydrogens will be returned
    Return a dict with the format {atom order : (H index, Neighbor Atom index)}
    '''
    ## find symmetrically equivalent hydrogens by canonical order
    ## if unique, no equivalent order will be given (unique rank for equivalent Hs)
    orders = list(Chem.CanonicalRankAtoms(mol, breakTies = not unique))
    ## find atom indices of hydrogens and the indices of their neighbor atoms
    H_neigh_ids = [(a.GetIdx(), a.GetNeighbors()[0].GetIdx()) 
                   for a in mol.GetAtoms() if a.GetSymbol() == 'H']
    ## return only one hydrogen per rank {order : (H index, Neighbor index)}
    return {orders[ids[0]] : ids for ids in H_neigh_ids}

def setTetherIdx(mol):
    mol2 = Chem.rdmolops.RemoveHs(mol)
    idx_list = [str(i) for i in range(1, mol2.GetNumAtoms()+1)]
    # set tethering atom ids for both versions of rdock
    mol.SetProp('rxdock.tethered_atoms',', '.join(idx_list))
    mol.SetProp('rdock.tethered_atoms',', '.join(idx_list))

def connectMols(lig, frag, id_ligH, id_ligNbr, id_fragH, id_fragNbr):
    """
    Connect two mols whose hydrogens are aligned into position
    Return a combined mol
    Function copied from here with modification
    https://github.com/molecularsets/moses/blob/master/moses/baselines/combinatorial.py
    """
    # align the fragament to the ligand 
    # fragment will move to the ligand to preserve docked poses
    _rmsd = AlignMol(frag, lig, 
                     atomMap=((id_fragH, id_ligNbr),(id_fragNbr, id_ligH)))
    # combine and make an editable mol object
    # ligand first, fragment second to preserve the atom idx of ligand
    combined = Chem.CombineMols(lig, frag)
    emol = Chem.EditableMol(combined)
    # form the bond
    emol.AddBond(id_ligNbr,
                 id_fragNbr + lig.GetNumAtoms(), 
                 # fragament atom idx increase by the num of atom in ligand
                 order=Chem.BondType.SINGLE)
    # remove the hydrogens from the ligand and the fragment
    emol.RemoveAtom(id_fragH + lig.GetNumAtoms())
    emol.RemoveAtom(id_ligH)
    # return the immutable mol object
    mol_out = emol.GetMol()
    Chem.SanitizeMol(mol_out)
    return mol_out

def functionalizeH(lig, frag, unique_H: bool = True, tether: bool = False ):
    '''
    Functionalize the ligand with a fragment on all unique positions of hydrogen atoms
    If unqiue flag is set to False, all hydrogen positions on the ligand will be used
    for forming the bonds. 
    Return a list of functionalized ligand
    '''
    # tether atom index set on the original ligand mol will be preserved when copied
    if tether: setTetherIdx(lig)
    fused_list = []
    # get combination of positions of hydrogens from both lig and frag
    # only unique hydrogens position will be used on the fragament
    combos = product(findHIds(lig, unique_H).values(), findHIds(frag, True).values())
    for combo in combos:
        id_ligH, id_ligNbr = combo[0]
        id_fragH, id_fragNbr = combo[1]
        # create a new copy (deep copy) of the mols
        cur_lig = copy.deepcopy(lig)
        cur_frag = copy.deepcopy(frag)
        fused = connectMols(cur_lig, cur_frag, id_ligH, id_ligNbr, id_fragH, id_fragNbr)
        fused_list.append(fused) 
    return fused_list
    
def _FuncHWrapper(args):
    return functionalizeH(*args[0], unique_H = args[1], tether=args[2])

def combinatorialFunc(lig_list, frag_list, propfilter, unique_H: bool = True, tether: bool = False):
    '''
    Enumerated all the functionalized molecules 
    from a list of ligand and a list of fragment
    Return a 3D list of lists with format 
    [[lig1-frag1, lig1-frag2, ..., lig1-fragN],
     [lig2-frag1, lig2-frag2, ..., lig2-fragN],
     ...
     [ligN-frag1, ligN-frag2, ..., ligN-fragN]]
     where lig_i-frag_j is a list of mols fused on various hydrogen positions
    '''
    args = list(zip(product(lig_list, frag_list), repeat(unique_H), repeat(tether)))
    fused = process_map(
            _FuncHWrapper,
            args,
            chunksize = 100
        )
    return propfilter.filterMols(fused, show_tqdm = False)

def _sdfFuncWrapper(args):
    '''
    Function wrapper for multiprocessing on functionalization and writing to file
    each thread processes one ligand and write to file
    '''
    lig_idx, lig, frags, path_template, propfilter, unique_H, tether = args
    # tether atom index set on the original ligand mol will be preserved when copied 
    if tether: setTetherIdx(lig)
    # get combination of positions of hydrogens from both lig and frag
    # only unique hydrogens position will be used on the fragament
    for frag_idx, frag in enumerate(frags):
        combos = product(findHIds(lig, unique_H).values(), findHIds(frag, True).values())
        fused_mols = []
        for combo in combos:
            id_ligH, id_ligNbr = combo[0]
            id_fragH, id_fragNbr = combo[1]
            # create a new copy (deep copy) of the mols
            cur_lig = copy.deepcopy(lig)
            cur_frag = copy.deepcopy(frag)
            fused_mols.append(connectMols(cur_lig, cur_frag, id_ligH, id_ligNbr, id_fragH, id_fragNbr))
        out_mols = propfilter.filterMols(fused_mols, show_tqdm = False)
        if len(out_mols) > 0:
            file_path = path_template.format(lig_idx, frag_idx)
            with open(file_path, 'w') as f:
                w = Chem.SDWriter(f)
                for m in out_mols:
                    w.write(m) 
                w.close()
        # else:
        #     print("Ligand {} and Fragment {} have no product that meets filtering criteria".format(lig_idx, frag_idx))

def sdfCombinatorialFunc(lig_path, frag_path, out_path, propfilter, unique_H: bool = True, tether: bool = False):
    '''
    Enumerated all the combinatorial of ligands and fragments on all H positions and functionalize the pairs
    Takes input from .sdf ligand file and .sdf fragment file 
    Files can contain multiple ligands or fragments to be connected
    Save the fused molecules in sdf files with presevation of ligands 3D coordinates
    Multithreaded: each thread processes one ligand and write to files
    '''
    if os.path.exists(out_path): 
        # Determined the file path and set file name template
        lig_name = os.path.split(lig_path)[-1].split('.')[0]
        out_file_template = lig_name+'_combined_{}_{}.sdf'
        out_path_template = os.path.join(out_path, out_file_template)
        
        ligs = [m for m in Chem.SDMolSupplier(lig_path, removeHs = False)]
        frags = [m for m in Chem.SDMolSupplier(frag_path, removeHs = False)]

        args = list((i, lig, frags, out_path_template, propfilter, unique_H, tether) for i, lig in enumerate(ligs))
        fused = process_map(
            _sdfFuncWrapper,
            args,
            chunksize = 1
        )
        print(len(ligs)*len(frags),"ligand-fragment pair processed.")
    else: 
        raise FileNotFoundError("Directory doesn't exist:", out_path)
        
if __name__ == "__main__":
    import sys
    import argparse

    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description=
'''
Ligand-Fragment Functionalization
DETAILS TO BE FILLED HERE

Available Filters: 
    SAS  - Sythetic Accessibility
    QED  - Quantitative Estimate of Druglikeness
    MW   - Molecular Weight
    LogP - Octanol-Water Partition Coefficient
    
    Filter criteria should start with relational signs, such as ">", "<", ">=", or "<=",
    followed by numerical values, and surrounded by quotation marks
    To specify both upper and lower bound, use two expressions separated by space.
    e.g, for Molecular Weight x: 300 < x < 500, use --MW "<500" ">300"

Usage Example:
DETAILS TO BE FILLED HERE
'''
                                    )
    # define extend action for argparse for python version lower than 3.8
    if sys.version_info.major == 3 and sys.version_info.minor < 8:
        class ExtendAction(argparse.Action):
            def __call__(self, parser, namespace, values, option_string=None):
                items = getattr(namespace, self.dest) or []
                items.extend(values)
                setattr(namespace, self.dest, items)
        parser.register('action', 'extend', ExtendAction)

    parser.add_argument('-l','--lig', type=str,
                        help='ligand .sdf file path')
    parser.add_argument('-f','--frag', type=str,
                        help='fragment .sdf file path')
    parser.add_argument('-o','--outpath', type=str, default="./out",
                        help='Directory path for file output. Default to the ./out')
    parser.add_argument('--allH', action='store_true', default=False,
                        help='''
                        Attach to all Hydrogen position instead of symmetrically unique ones. \n
                        Without this flag, it only replace symmetrically unique hydrogens
                        ''')
    parser.add_argument('--tether', action='store_true', default=False,
                        help='''
                        Set the tether docking atom ids for the functionalized ligand.\n
                        If the flag is present, the atoms from the ligand will be fixed during docking, 
                        and only the fragment parts are allowed to move freely.\n
                        Default to no tethering (without this flag)
                        ''')
    # options for property filters 
    parser.add_argument('--SAS', '--sas', action="extend", nargs="+", 
                        help='Sythetic Accessibility')
    parser.add_argument('--QED', '--qed', action="extend", nargs="+", type=str,
                        help='Quantitative Estimate of Druglikeness')
    parser.add_argument('--MolWeight', '--MW', '--mw', action="extend", nargs="+", type=str,
                        help='Molecular Weight')
    parser.add_argument('--LogP', '--logP', '--logp', action="extend", nargs="+", type=str,
                        help='Octanol-Water Partition Coefficient')

    args = parser.parse_args()
    
    if not os.path.exists(args.outpath):
        os.makedirs(args.outpath)
        print('Directory Made:',args.outpath)
        
    pf = PropFilter(args.__dict__)
    if len(pf.filter_expr) == 0:
        print("No filter set for output mols")
    else:
        print('Filters set as below:')
        print(pf.filter_expr)

    uniqueH = not args.allH
    sdfCombinatorialFunc(args.lig, args.frag, args.outpath, pf, uniqueH, args.tether)